wayland-client/src/globals.rs

```diff
@@ -0,0 +1,291 @@
+//! At the startup of your Wayland app, the initial step is generally to retrieve the list of globals
+//! advertized by the compositor from the registry. Using the [`Dispatch`] mechanism for this task can be
+//! very unpractical, this is why this module provides a special helper for handling the registry.
+//!
+//! The entry point of this helper is the [`registry_queue_init`] function. Given a reference to your
+//! [`Connection`] it will create an [`EventQueue`], retrieve the initial list of globals, and register a
+//! handler using your provided `Dispatch<WlRegistry,_>` implementation for handling dynamic registry events.
+//!
+//! ## Example
+//!
+//! ```no_run
+//! use wayland_client::{
+//!     Connection, Dispatch, QueueHandle,
+//!     globals::{registry_queue_init, Global},
+//!     protocol::{wl_registry, wl_compositor},
+//! };
+//! # use std::sync::Mutex;
+//! # struct State;
+//!
+//! // You need to provide a Dispatch<WlRegistry, Mutex<Vec<Global>>> impl for your app
+//! impl wayland_client::Dispatch<wl_registry::WlRegistry, Mutex<Vec<Global>>> for State {
+//!     fn event(
+//!         state: &mut State,
+//!         proxy: &wl_registry::WlRegistry,
+//!         event: wl_registry::Event,
+//!         // This mutex contains an up-to-date list of the currently known globals
+//!         // including the one that was just added or destroyed
+//!         data: &Mutex<Vec<Global>>,
+//!         conn: &Connection,
+//!         qhandle: &QueueHandle<State>,
+//!     ) {
+//!         /* react to dynamic global events here */
+//!     }
+//! }
+//!
+//! let conn = Connection::connect_to_env().unwrap();
+//! let (globals, queue) = registry_queue_init::<State>(&conn).unwrap();
+//!
+//! # impl wayland_client::Dispatch<wl_compositor::WlCompositor, ()> for State {
+//! #     fn event(
+//! #         state: &mut State,
+//! #         proxy: &wl_compositor::WlCompositor,
+//! #         event: wl_compositor::Event,
+//! #         data: &(),
+//! #         conn: &Connection,
+//! #         qhandle: &QueueHandle<State>,
+//! #     ) {}
+//! # }
+//! // now you can bind the globals you need for your app
+//! let compositor: wl_compositor::WlCompositor = globals.bind(&queue.handle(), 4..=5, ()).unwrap();
+//! ```
+
+use std::{
+    ops::RangeInclusive,
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc, Mutex,
+    },
+};
+
+use wayland_backend::{
+    client::{Backend, InvalidId, ObjectData, ObjectId, WaylandError},
+    protocol::Message,
+};
+
+use crate::{
+    protocol::{wl_display, wl_registry},
+    Connection, Dispatch, EventQueue, Proxy, QueueHandle,
+};
+
+/// Initialize a new event queue with its associated registry and retrieve the initial list of globals
+///
+/// See [the module level documentation](self) for more.
+pub fn registry_queue_init<State>(
+    conn: &Connection,
+) -> Result<(GlobalList, EventQueue<State>), GlobalError>
+where
+    State: Dispatch<wl_registry::WlRegistry, Mutex<Vec<Global>>> + 'static,
```

Comment:
Feels a bit raw to expose `Mutex<Vec<Global>>` as the user data type. Do we want to provide a wrapper type that is a little nicer to work with?

Reply:
I did wonder about that, but didn't have the motivation to do it before having your first opinion over the overall API. Will do. :+1: